### PR TITLE
Offer default settings and a method to update it.

### PR DESCRIPTION
**Pull request by core1024**

### CODE REVIEW

File: src/state_types/models/context.rs [hunk 1]
```diff
@@ -28,12 +29,28 @@ pub struct Auth {
 pub struct CtxContent {
     pub auth: Option<Auth>,
     pub addons: Vec<Descriptor>,
+    pub settings: HashMap<String, String>,
 }
 impl Default for CtxContent {
     fn default() -> Self {
+            let mut settings = HashMap::new();
```
Comment (Ivshti) [inline src/state_types/models/context.rs]: you should use `maplit` or `vec![("language", "eng"), ...].iter().cloned().collect()`

File: src/state_types/models/context.rs [hunk 2]
```diff
@@ -198,10 +221,15 @@ fn authenticate<Env: Environment + 'static>(action: ActionUser, req: APIRequest)
...
             };
+
+            // This is used only for authenticated users.
+            let settings = HashMap::new();
+
             api_fetch::<Env, CollectionResponse, _>(pull_req).map(
                 move |CollectionResponse { addons, .. }| CtxContent {
                     auth: Some(Auth { key, user }),
                     addons,
+                    settings,
```
Comment (Ivshti) [inline src/state_types/models/context.rs]: just inline `settings: Default::default()`

File: src/state_types/models/context.rs [hunk 3]
```diff
@@ -85,6 +102,12 @@ impl<Env: Environment + 'static> Update for Ctx<Env> {
                 self.content.addons.push(*descriptor.to_owned());
                 Effects::one(save_storage::<Env>(&self.content))
             }
+            Msg::Action(Action::Settings(ActionSettings::Store(settings))) => {
```
Comment (Ivshti) [inline src/state_types/models/context.rs]: not sure I like this - how about just passing the entire `settings` struct every time, so you would just `*self.content.settings = settings`

File: src/state_types/models/settings.rs [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+#[serde(rename_all = "camelCase")]
+pub enum SsProfileName {
+    Default,
+    Soft,
+    Fast,
+    Custom,
+}
+impl SsProfileName {
+    fn default_profile() -> Self {
+        SsProfileName::Default
```
[the remark this answers is not in the document]
Comment (core1024) [reply src/state_types/models/settings.rs]: Default implemented

File: src/state_types/models/settings.rs [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+    pub use_external_player: String,
+    // We can't override Esc in browser so this option is pointless here
+    // pub player_esc_exits_fullscreen:  String,
+    pub pause_on_lost_focus: String,
+    pub show_vid_overview: String,
+}
+
+impl Settings {
+    fn get_endpoint(&self) -> String {
+        format!("{}{}", self.server_url, "settings")
```
[the remark this answers is not in the document]
Comment (core1024) [reply src/state_types/models/settings.rs]: Now using path

File: src/state_types/models/settings.rs [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+        // web_sys::console::log_1(&format!("Update Settings!").into());
+        match msg {
+            // This is triggered after loading the settings from local storage
+            Msg::Internal(CtxLoaded(_))
+            | Msg::Action(Action::Settings(ActionSettings::LoadStreamingServer)) => {
+                web_sys::console::log_1(&"Load Ss Settings!".to_string().into());
+                let url = &ctx.content.settings.get_endpoint();
+                match Request::get(url).body(()) {
+                    Ok(resp) => Effects::one(Box::new(
+                        Env::fetch_serde::<_, SsSettings>(resp)
```
[the remark this answers is not in the document]
Comment (core1024) [reply src/state_types/models/settings.rs]: Done

**Pull request merged**